### PR TITLE
Add deferrable mode to Kubernetes and GKE operators

### PR DESCRIPTION
**Pull request by MrGeorgeOwl**
---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Comment by dgouju** [comment]
Hello @MrGeorgeOwl ,  quick comment on GKEStartPodOperator __init__():
Missing __init__() missing deferrable: bool = False,
Missing self.deferrable = deferrable
?

File: airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py [hunk 1]
```diff
@@ -272,6 +276,8 @@ def __init__(
         pod_runtime_info_envs: list[k8s.V1EnvVar] | None = None,
         termination_grace_period: int | None = None,
         configmaps: list[str] | None = None,
+        deferrable: bool = False,
+        poll_interval: float = 2,
```
Comment (kaxil) [inline airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py:285]: What does `poll_interval` mean for non-async operator?

File: airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py [hunk 2]
```diff
@@ -437,6 +442,10 @@ def hook(self) -> KubernetesHook:
     def client(self) -> CoreV1Api:
         return self.hook.core_v1_client
 
+    def get_hook(self):
+        warnings.warn("get_hook is deprecated. Please use hook instead.", DeprecationWarning, stacklevel=2)
+        return self.hook
+
```
Comment (kaxil) [inline airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py:451]: Why this unrelated change of moving the hook position?

File: docs/apache-airflow-providers-google/operators/cloud/kubernetes_engine.rst [hunk 3]
```diff
@@ -131,6 +131,14 @@ And then use it in other operators:
     :start-after: [START howto_operator_gke_xcom_result]
     :end-before: [END howto_operator_gke_xcom_result]
 
+You can use deferrable mode for this action in order to run the operator asynchronously:
```
Comment (kaxil) [inline docs/apache-airflow-providers-google/operators/cloud/kubernetes_engine.rst:134]: There are some limitations of using async KPO in its current state?

What are they? let's document it

File: airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py (hunk 1 above)
Comment (MrGeorgeOwl) [reply airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py:285]: For non-async nothing but it will give user opportunity to configure how often trigger should check pod operator status if deferrable mode is on

File: airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py (hunk 2 above)
Comment (MrGeorgeOwl) [reply airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py:451]: It is little refactoring for improving code readability so all cached methods and hook properties can be found in the same place

File: docs/apache-airflow-providers-google/operators/cloud/kubernetes_engine.rst (hunk 3 above)
Comment (MrGeorgeOwl) [reply docs/apache-airflow-providers-google/operators/cloud/kubernetes_engine.rst:134]: Could you be more descriptive about the idea you propose? I am not sure what you meant. That string is made for docs reader to inform them that deferrable mode is implemented for that pattern of usage

File: airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py (hunk 1 above)
Comment (kaxil) [reply airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py:285]: Please update docstring accordingly on L212 to say it is only for async

File: airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py (hunk 2 above)
Comment (kaxil) [reply airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py:451]: Let's avoid those unrelated changes please -- makes it difficult for checking git history

File: airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py [hunk 4]
```diff
@@ -484,15 +496,21 @@ def extract_xcom(self, pod: k8s.V1Pod):
             return json.loads(result)
 
     def execute(self, context: Context):
-        remote_pod = None
+        """Based on the deferrable parameter runs the pod asynchronously or synchronously"""
+        if self.deferrable:
+            self.asynchronously(context)
+        else:
+            return self.synchronously(context)
```
Comment (kaxil) [inline airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py]: Can you rename this too.

@VladaZakharova Apologies about delay in our contribution - Could you rebase once https://github.com/apache/airflow/pull/28523 is merged. That PR also has additional feature of ping-pong to show periodic logs which our customers found very useful

File: airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py [hunk 5]
```diff
@@ -509,48 +527,141 @@ def execute(self, context: Context):
...
         ti.xcom_push(key="pod_namespace", value=self.pod.metadata.namespace)
         if self.do_xcom_push:
             return result
 
-    def _read_pod_log_events(self, pod, *, reraise=True):
-        """Will fetch and emit events from pod"""
-        with _optionally_suppress(reraise=reraise):
-            for event in self.pod_manager.read_pod_events(pod).items:
-                self.log.error("Pod Event: %s - %s", event.reason, event.message)
+    def asynchronously(self, context: Context):
```
Comment (kaxil) [inline airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py]: `execute_async` would be a better name.

File: airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py (hunk 4 above)
Comment (VladaZakharova) [reply airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py:505]: We have also implemented some features that were and still are very urgent for our costumer. That is why it was important to reach you and your team earlier, to discuss these changes. So, may be after our current PR will be merged you can rebase and add additional logic.

Comment (kaxil) [reply airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py]: If you prefer you contribute first and @phanikumv rebases his PR -- that is fine too

**Comment by kaxil** [comment]
https://github.com/apache/airflow/pull/28523 was created first, let's rebase on top of it and add any features if needed once it is merged.

File: airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py (hunk 4 above)
Comment (kaxil) [reply airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py]: https://github.com/apache/airflow/pull/28523#issuecomment-1362176160 -- Let's get this PR merged first since it was created first. We will rebase that PR on main

File: airflow/providers/cncf/kubernetes/hooks/kubernetes.py [hunk 6]
```diff
@@ -406,3 +418,83 @@ def _get_bool(val) -> bool | None:
...
+    """Hook to use Kubernetes SDK asynchronously."""
+
+    def __init__(self, config_dict: dict | None = None, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.config_dict = config_dict
+
+    async def _load_config(self):
+        """Returns Kubernetes API session for use with requests"""
+        in_cluster = self._coalesce_param(self.in_cluster, self._get_field("in_cluster"))
+        cluster_context = self._coalesce_param(self.cluster_context, self._get_field("cluster_context"))
```
Comment (kaxil) [inline airflow/providers/cncf/kubernetes/hooks/kubernetes.py]: `self._get_field` calls `self.conn_extras` which has a DB call. 

Can you override it please so that we can make it an async call

File: airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py [hunk 7]
```diff
@@ -484,15 +498,21 @@ def extract_xcom(self, pod: k8s.V1Pod):
...
 
     def execute(self, context: Context):
-        remote_pod = None
+        """Based on the deferrable parameter runs the pod asynchronously or synchronously"""
+        if self.deferrable:
+            self.asynchronously(context)
+        else:
+            return self.synchronously(context)
+
+    def synchronously(self, context: Context):
```
Comment (kaxil) [inline airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py]: same as https://github.com/apache/airflow/pull/28230/files#r1054839109

`synchronously` name for a method alone doesn't make much sense

**Review by kaxil — COMMENTED** [review]
If you can separate this into 2 PRs - one for async mode for KPO and the other for async mode for GKE, that will help in reviewing

File: airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py (hunk 2 above)
Comment (VladaZakharova) [reply airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py:451]: We have moved this method to be placed in one place as other methods after all the cached_properties. We will avoid such kind of changes in next PRs, thanks!

File: airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py (hunk 4 above)
Comment (VladaZakharova) [reply airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py]: Sure, would be nice of you. We have added some changes to prevent memory leak and are working now on other changes to improve performance for the operator. We will let you know when it will be ready :)

**Comment by VladaZakharova** [comment]
> 

As we are adding new feature to existing operator instead of creating new operator for deferrable mode, we decided to also use KubernetesPodOperator as a base class for the GKEStartPodOperator to instantiate needed behavior. So splitting this to 2 separate PRs will be not an option here as we will need to wait until KubernetesPodOperator will be merged to main and only then add new PR for the GKEStartPodOperator.

**Comment by kaxil** [comment]
> > 
> 
> As we are adding new feature to existing operator instead of creating new operator for deferrable mode, we decided to also use KubernetesPodOperator as a base class for the GKEStartPodOperator to instantiate needed behavior. So splitting this to 2 separate PRs will be not an option here as we will also need to wait until KubernetesPodOperator will be merged to main and only then add new PR for the GKEStartPodOperator.

Yup, that is exactly what I suggest, though. If the PRs are big, it will take more time for committers like myself to review and merge. If the PRs are broken down into smaller chunks it is easier to review and merge faster. What I would suggest:

1) Convert this PR to `Add "deferrable" mode to KPO`, and we will get this merged 
2) Once (1) is merged, create a separate PR for adding deferrable mode to GKE

**Comment by VladaZakharova** [comment]
@eladkal @potiuk @kaxil 
Hi Team! :)
Could we please review the changes in this PR?

File: generated/provider_dependencies.json [hunk 8]
```diff
@@ -209,8 +209,10 @@
   "cncf.kubernetes": {
     "deps": [
       "apache-airflow>=2.3.0",
+      "asgiref>=3.5.2",
       "cryptography>=2.0.0",
-      "kubernetes>=21.7.0,<24"
+      "kubernetes>=21.7.0,<24",
+      "kubernetes_asyncio==24.2.2"
```
Comment (kaxil) [inline generated/provider_dependencies.json]: Let's not pin to a specific version, Airflow is also used as a library and an application. Pinning it to a specific version restricts what deps can be used with Airflow

**Comment by kaxil** [comment]
> @eladkal @potiuk @kaxil Hi Team! :) Could we please review the changes in this PR?

As I had mentioned in https://github.com/apache/airflow/pull/28230#issuecomment-1362785141 , if the PR is big, it will take time to review and makes our life as reviewers hard. So please assume it will take some time to fully review the PR

File: airflow/providers/google/cloud/operators/kubernetes_engine.py [hunk 9]
```diff
@@ -305,11 +303,9 @@ def __init__(
         *,
         location: str,
         cluster_name: str,
-        use_internal_ip: bool = False,
```
Comment (kaxil) [inline airflow/providers/google/cloud/operators/kubernetes_engine.py:308]: This is going to break backwards compat!

File: airflow/providers/google/cloud/operators/kubernetes_engine.py [hunk 10]
```diff
@@ -305,11 +303,9 @@ def __init__(
...
         project_id: str | None = None,
         gcp_conn_id: str = "google_cloud_default",
         impersonation_chain: str | Sequence[str] | None = None,
-        regional: bool = False,
```
Comment (kaxil) [inline airflow/providers/google/cloud/operators/kubernetes_engine.py:312]: This is going to break backwards compat!

File: airflow/providers/google/cloud/operators/kubernetes_engine.py (hunk 9 above)
Comment (MrGeorgeOwl) [reply airflow/providers/google/cloud/operators/kubernetes_engine.py:308]: Replace it with optional parameter and add deprecation message

File: airflow/providers/google/cloud/operators/kubernetes_engine.py (hunk 10 above)
Comment (MrGeorgeOwl) [reply airflow/providers/google/cloud/operators/kubernetes_engine.py:312]: Replace it with optional parameter and add deprecation message

File: generated/provider_dependencies.json (hunk 8 above)
Comment (MrGeorgeOwl) [reply generated/provider_dependencies.json]: Change it to range

**Comment by VladaZakharova** [comment]
@kaxil @MrGeorgeOwl @lwyszomi 
Hi Team! 
Please, check separate PR for the KubernetesPodOperator: https://github.com/apache/airflow/pull/29017
Hope that it can really speed up little bit the review process :)

**Comment by eladkal** [comment]
@VladaZakharova can you rebase and fix conflicts?

**Comment by VladaZakharova** [comment]
@eladkal 
Hi!
This PR was splitted to 2 separate. Please, check the link in the comment above :)
I think this PR can be closed in this case. But i am not the owner, so i cant.
